### PR TITLE
Pytorch & ONNX Clean up & Improvement

### PR DESCRIPTION
**Pull request by Duchstf**
In this PR:

* Cleaned up and debug ONNX & Pytorch codes.
* Organize ONNX & Pytorch according to Keras scheme.
* Initial support for ONNX & Pytorch from Python API. 

The codes were tested and it worked with some of my test models (MLP 2 layers/3layers). Due to problems with CNN backend (no support for channel_first), I didn't test it on any CNN models. 

@thesps @vloncar Let me know how I should proceed and if I need to add more tests. 

Thanks!

### CODE REVIEW

File: hls4ml/converters/onnx/pooling.py
Comment (vloncar) [inline hls4ml/converters/onnx/pooling.py:23]: Same as with convolution, computing output shape should be shared by all parsers

File: hls4ml/converters/onnx_to_hls.py
Comment (vloncar) [inline hls4ml/converters/onnx_to_hls.py:11]: Docstrings instead of ASCII art comments

File: hls4ml/converters/onnx_to_hls.py
Comment (vloncar) [inline hls4ml/converters/onnx_to_hls.py:120]: Why is this not implemented as a regular hls4ml optimizer? `Transpose` is a valid layer and should appear in the internal model, then if it is placed before and after a real layer (which only happens when using one very poorly designed ONNX conversion tool) we can optimize it away.

File: hls4ml/converters/onnx_to_hls.py
Comment (vloncar) [inline hls4ml/converters/onnx_to_hls.py:187]: I would prefer to just have `OnnxModel` and then figure out how to proceed based on the object that is in the config (a real model or a string that is its path on the filesystem).

File: hls4ml/converters/onnx_to_hls.py
Comment (vloncar) [inline hls4ml/converters/onnx_to_hls.py:202]: Why assuming just one? You can get the list of input layers from the onnx graph and create one input layer for each as you parse.

File: hls4ml/converters/onnx_to_hls.py
Comment (vloncar) [inline hls4ml/converters/onnx_to_hls.py:216]: Reshape should be handled, we have models that need to handle it, more will come in the future. Squeeze/Unsqueeze can probably be folded into Reshape as well.

File: hls4ml/converters/onnx_to_hls.py
Comment (vloncar) [inline hls4ml/converters/onnx_to_hls.py:276]: If you don't pass input/output layers only sequential models are supported. This should be solved now or documented as a limitation in the docs for future improvement

File: hls4ml/converters/pytorch/convolution.py
Comment (vloncar) [inline hls4ml/converters/pytorch/convolution.py:32]: Same as mentioned in the ONNX converter, this can be made into an utility function.

File: hls4ml/converters/pytorch/convolution.py
Comment (vloncar) [inline hls4ml/converters/pytorch/convolution.py:71]: Ditto

File: hls4ml/converters/pytorch_to_hls.py
Comment (vloncar) [inline hls4ml/converters/pytorch_to_hls.py:12]: Docstrings instead of comments

File: hls4ml/converters/pytorch_to_hls.py
Comment (vloncar) [inline hls4ml/converters/pytorch_to_hls.py:100]: Use just `PyTorchModel`

File: hls4ml/converters/pytorch_to_hls.py
Comment (vloncar) [inline hls4ml/converters/pytorch_to_hls.py:186]: Same as with ONNX, this should have input and output layers

File: hls4ml/utils/config.py
Comment (vloncar) [inline hls4ml/utils/config.py:220]: Docstring here

File: setup.py
Comment (vloncar) [inline setup.py:38]: How is this cleaner code? This is the last nitpick :-)

**Review by vloncar — CHANGES_REQUESTED** [review]
Thanks Duc, awesome work. Minor stuff only. There aren't many things to change, the comments are duplicated between PyTorch and ONNX converters

File: setup.py
Comment (Duchstf) [reply setup.py:38]: Sorry I added a dependency & then remove it. Didn't realize it changed the code.

File: hls4ml/converters/onnx_to_hls.py
Comment (Duchstf) [reply hls4ml/converters/onnx_to_hls.py:120]: Sorry, I'm not sure what you mean by this? so do suggest that I implement it [here](https://github.com/fastmachinelearning/hls4ml/tree/master/hls4ml/model/optimizer) instead of an onnx optimizer?

Comment (vloncar) [reply hls4ml/converters/onnx_to_hls.py:139]: What ONNX models have this issue? PyTorch or Keras? If the issue is with PyTorch, does [this](https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html) help?

Performing graph transformations in the converter will make it harder to maintain it in the future. We've been there before. For example, your current transformation is dangerous, you need to make extra checks to ensure that the axes of transpose match (meaning you are transposing back into original format). Also how would you add future optimizations on top of this? You're modifying the graph as you iterate it, that's asking for trouble.

File: hls4ml/utils/config.py
Comment (jmduarte) [inline hls4ml/utils/config.py:246]: Remove commented code

File: hls4ml/utils/config.py
Comment (jmduarte) [inline hls4ml/utils/config.py:230]: Remove commented code

File: hls4ml/converters/keras/pooling.py
Comment (jmduarte) [inline hls4ml/converters/keras/pooling.py:4]: No star import

**Review by jmduarte — CHANGES_REQUESTED** [review]

File: hls4ml/utils/config.py
Comment (Duchstf) [reply hls4ml/utils/config.py:221]: This makes perfect sense! However I would plan to include this with the quantization PR (which will require us to generate the config with higher granularity).

The purpose of this PR is mainly to set up a basic software framework that we can develop on later.

**Comment by vloncar** [comment]
@Duchstf Is this ready for a new review? Can you resolve the conflicts?

**Comment by Duchstf** [comment]
@vloncar Thanks for checking in! I was actually about to message you! 

I addressed most of the comments, but I'm still kind of stuck at squeeze/unsqeeze and transpose layer. Is there any implementation of these layers in the Keras converter?

After this I think I'll need to move the "double transpose" optimization to hls model graph, and then add in input/output layers for onnx.

**Comment by vloncar** [comment]
Hi Duc,

There is [`Transpose`](https://github.com/fastmachinelearning/hls4ml/blob/master/hls4ml/model/hls_layers.py#L1486) layer in hls4ml, your converters should use that, and then you should write an optimizer to get rid of the double transpose. Squeeze/Unsqueeze can be mapped to `Reshape`. I don't understand the input/output layers for ONNX, can you elaborate?

**Comment by Duchstf** [comment]
@vloncar Thank you! I'll try to work on this.

Regarding the input/outputs, there is some inconsistency in ONNX that I need to resolve. For example, this is how the output layer is obtained in onnx in our converter:

```
output_layers = [x.name for x in model.graph.output]
```

If I have this model for example:

Screen Shot 2021-03-30 at 1 41 11 PM

Then `output_layers` would be `["36"]`, however, the correct output layer is named `"Transpose_2"` in this case. So if you try to access this layer in the hls model graph it would not be available. 

Another problem in this case is that the transpose layer is optimized away, so I need to remap the output layer to the Softmax layer.

**Comment by Duchstf** [comment]
@jmduarte @vloncar Hello, sorry to be late on this!! 

I addressed all the major comments now, and I think the current code base provide a nice foundation to be extended in the future. The code has been tested through a [rigorous test bench that Javier wrote](https://github.com/jmduarte/pytorch_dev_hls4ml/blob/main/MLP_performance.py) for the 3-layer MLP jet-tagging model. 

With the current codes ONNX can also support more complicated non-sequential architecture, whereas Pytorch can only process linear models for now (noted in the doc). More plans on other additions to come. 

I'm requesting that this PR is merged now to prepare for an up-coming GNN PR of Javier's student. I will also work on later extensions after I can create more example ONNX models through the QONNX frontend.

Thanks!

**Review by jmduarte — APPROVED** [review]

**Comment by jmduarte** [comment]
The changes look good to me.

I've tested this branch with the 3-layer jet classification model and find perfect agreement between the performance for the ONNX-converted and PyTorch-converted hls4ml models.

From here: https://github.com/jmduarte/pytorch_dev_hls4ml/blob/main/MLP_performance.py

```
PyTorch Accuracy: 0.7705041666666667
PyTorch j_g eff_bkg @ eff_sig=0.5: 0.01709535431921063
PyTorch j_g auc:                   0.9428119136939714
PyTorch j_q eff_bkg @ eff_sig=0.5: 0.04826059063347199
PyTorch j_q auc:                   0.9077707675852923
PyTorch j_w eff_bkg @ eff_sig=0.5: 0.0043961201501877345
PyTorch j_w auc:                   0.955058513686781
PyTorch j_z eff_bkg @ eff_sig=0.5: 0.0013408182643446685
PyTorch j_z auc:                   0.9506058228986847
PyTorch j_t eff_bkg @ eff_sig=0.5: 0.0036755876241293976
PyTorch j_t auc:                   0.9636456150795238
PyTorch average eff_bkg @ eff_sig=0.5: 0.004639871231215533
PyTorch average auc                  : 0.9439785265888506
PyTorch->hls4ml ap_fixed<16,6> Accuracy: 0.7500875
PyTorch->hls4ml j_g eff_bkg @ eff_sig=0.5: 0.01775768195756931
PyTorch->hls4ml j_g auc:                   0.9406825048141373
PyTorch->hls4ml j_q eff_bkg @ eff_sig=0.5: 0.048664141884480865
PyTorch->hls4ml j_q auc:                   0.9064459952657805
PyTorch->hls4ml j_w eff_bkg @ eff_sig=0.5: 0.004151022110972048
PyTorch->hls4ml j_w auc:                   0.9530614087511615
PyTorch->hls4ml j_z eff_bkg @ eff_sig=0.5: 0.0021025282510930016
PyTorch->hls4ml j_z auc:                   0.9489382343086378
PyTorch->hls4ml j_t eff_bkg @ eff_sig=0.5: 0.0031534871093382897
PyTorch->hls4ml j_t auc:                   0.9621728642823041
PyTorch->hls4ml average eff_bkg @ eff_sig=0.5: 0.004554429724725144
PyTorch->hls4ml average auc                  : 0.9422602014844041
```
```
ONNX Accuracy: 0.7705916666666667
ONNX j_g eff_bkg @ eff_sig=0.5: 0.01714750610175856
ONNX j_g auc:                   0.9428539660674484
ONNX j_q eff_bkg @ eff_sig=0.5: 0.048286459303408455
ONNX j_q auc:                   0.907730672199003
ONNX j_w eff_bkg @ eff_sig=0.5: 0.0043961201501877345
ONNX j_w auc:                   0.9550627050807532
ONNX j_z eff_bkg @ eff_sig=0.5: 0.0013095151142043261
ONNX j_z auc:                   0.9506387713869444
ONNX j_t eff_bkg @ eff_sig=0.5: 0.003790449737383441
ONNX j_t auc:                   0.963492754765287
ONNX average eff_bkg @ eff_sig=0.5: 0.004619214586735845
ONNX average auc                  : 0.9439557738998872
ONNX->hls4ml Accuracy: 0.7500875
ONNX->hls4ml j_g eff_bkg @ eff_sig=0.5: 0.01775768195756931
ONNX->hls4ml j_g auc:                   0.9406825048141373
ONNX->hls4ml j_q eff_bkg @ eff_sig=0.5: 0.048664141884480865
ONNX->hls4ml j_q auc:                   0.9064459952657805
ONNX->hls4ml j_w eff_bkg @ eff_sig=0.5: 0.004151022110972048
ONNX->hls4ml j_w auc:                   0.9530614087511615
ONNX->hls4ml j_z eff_bkg @ eff_sig=0.5: 0.0021025282510930016
ONNX->hls4ml j_z auc:                   0.9489382343086378
ONNX->hls4ml j_t eff_bkg @ eff_sig=0.5: 0.0031534871093382897
ONNX->hls4ml j_t auc:                   0.9621728642823041
ONNX->hls4ml average eff_bkg @ eff_sig=0.5: 0.004554429724725144
ONNX->hls4ml average auc                  : 0.9422602014844041
```

**Comment by jmduarte** [comment]
@Duchstf can you fix the minor conflict we introduced?

@vloncar can we merge this after that's fixed? a follow-up PR should be to add testing to the test suite.

**Comment by Duchstf** [comment]
@jmduarte there are some conflicts from rebasing so I just resolve the easier conflict from merging ... hopefully this doesn't introduce much pain for Abdul ...

**Comment by vloncar** [comment]
I tried dogfooding the onnx converter with a model produced from qonnx project (but without the `Quant` nodes)
[test_285.zip](https://github.com/fastmachinelearning/hls4ml/files/6902912/test_285.zip)
The model is very simple (try it out in [Netron](https://netron.app/)), it should be easily parsable.

**Comment by Duchstf** [comment]
@vloncar Thanks for the model, yes it shouldn't be too hard to fix!

**Comment by Duchstf** [comment]
@vloncar I made some changes and it was able to parse all the layers, this is the current output:

```
Interpreting Model ...
Output layers:  ['Softmax']
Input shape: [None, 28, 28, 1]
Topology:
Layer name: Transpose4, layer type: Transpose, current shape: [[None, 28, 28, 1]]
Layer name: conv2d_0_m, layer type: Conv2D, current shape: [[None, 1, 28, 28]]
Layer name: Relu2, layer type: Activation, current shape: [[None, 32, 14, 14]]
Layer name: conv2d_1_m, layer type: Conv2D, current shape: [[None, 32, 14, 14]]
Layer name: Relu1, layer type: Activation, current shape: [[None, 64, 6, 6]]
Layer name: conv2d_2_m, layer type: Conv2D, current shape: [[None, 64, 6, 6]]
Layer name: Relu, layer type: Activation, current shape: [[None, 64, 3, 3]]
Layer name: PushTranspose_0, layer type: Transpose, current shape: [[None, 64, 3, 3]]
Layer name: dense, layer type: Dense, current shape: [[None, 576]]
Layer name: Add, layer type: Merge, current shape: [[None, 10]]
Layer name: Softmax, layer type: Softmax, current shape: [[None, 10]]
Creating HLS model
[0, 3, 1, 2]
Traceback (most recent call last):
  File "qonnx_test.py", line 5, in <module>
    hls_model = hls4ml.converters.convert_from_onnx_model(onnx_model, output_dir='output_hls')
  File "/data/dhoang/onnx_merge/hls4ml/hls4ml/converters/__init__.py", line 325, in convert_from_onnx_model
    return onnx_to_hls(config)
  File "/data/dhoang/onnx_merge/hls4ml/hls4ml/converters/onnx_to_hls.py", line 305, in onnx_to_hls
    hls_model = HLSModel(config, reader, layer_list, input_layers, output_layers)
  File "/data/dhoang/onnx_merge/hls4ml/hls4ml/model/hls_model.py", line 317, in __init__
    self._make_graph(layer_list)
  File "/data/dhoang/onnx_merge/hls4ml/hls4ml/model/hls_model.py", line 332, in _make_graph
    self.graph[name] = self.make_node(kind, name, layer, inputs, outputs)
  File "/data/dhoang/onnx_merge/hls4ml/hls4ml/model/hls_model.py", line 363, in make_node
    node = layer_map[kind](self, name, attributes, inputs, outputs)
  File "/data/dhoang/onnx_merge/hls4ml/hls4ml/model/hls_layers.py", line 375, in __init__
    self.initialize()
  File "/data/dhoang/onnx_merge/hls4ml/hls4ml/model/hls_layers.py", line 1519, in initialize
    raise Exception('ERROR: Transpose of tensors with rank > 3 is not yet supported.')
Exception: ERROR: Transpose of tensors with rank > 3 is not yet supported.
```

`[0, 3, 1, 2]` here is the print out of the transpose layer's `perm`, I know that the transpose layers only operate on the conv's input/output, so I'm not sure whether to fix this or throwing an error in this case.

**Comment by jmduarte** [comment]
I think the problem is that the current HLS Transpose layer is written with the `Keras` `perm` option in mind: 

https://www.tensorflow.org/api_docs/python/tf/keras/layers/Permute#args

which for the same use-case would be written like this: `[3, 1, 2]`, i.e. they exclude the batch/samples dimension (that is not ever transposed with anything).

In ONNX, they explicitly include the batch/samples dimension, so the `perm` array is written like this `[0, 3, 1, 2]`.

So maybe we should just remove the first entry of the `perm` array in the ONNX converter before passing it to the `Transpose` HLS layer?

**Comment by Duchstf** [comment]
@vloncar @jmduarte I pushed some small fixes to support parsing this model, everything parses fine now and the results of the 3-layer test was not affected!

Here are some detailed changes:

- Remove some bugs in Conv layer parsing.
- Extend the merge layer parsing to recognize when an Add layer is `AddBias` and use the current infrastructure in tf to merge it with `MatMul` ( which was parsed to `Dense`).
- Extend the reader to find the correct bias weights in `AddBias.`
- Ignore the batch dimension in `Transpose`.
- Some extensions to make the codes more efficient.

**Comment by vloncar** [comment]
This needs a far more thorough testing to be considered ready. For example, just adding a simple `MaxPool` layer to the model above again breaks the conversion.

I've created [test case](https://gist.github.com/vloncar/dd89536ed36cf3b55eea93506d20be55) with some nonsense models created with Keras. I think this should (reasonably) pass before we merge this PR. The similar thing should be done for ONNX models exported from PyTorch.

**Comment by Duchstf** [comment]
@vloncar @jmduarte So here is the summary of the testing that I did:

1. For the keras2onnx test, the converter can parse all the models (need to remove the Clip in the 2d layer but I think that's a keras2onnx bug), I also tried to vary different architecture and they were parsed just fine. For compilation, the branch model compiled, but not the 1d/2d model. I think this is mainly a backend problem (channel first operations not supported) and not the converter. Additionally, I also manually checked all the output shapes to make sure they are consistent. 

2. I made a similar testing script but converting from pytorch models, in my script, I can compile my test2d and branch model, but compilation for 1d model also failed because of a similar backend problem.

Overall, I think the converter is in good shape and is ready to be merged now, I can take these testing scripts, refine them and do a follow-up PR on testing (maybe after the Keras testing infrastructure is integrated). 

For reproducibility, I put the testing scripts and well as the conda environment setup[ in this repo](https://github.com/Duchstf/onnx_test).

**Comment by vloncar** [comment]
`Clip` is not a bug. ReLU is really a `Clip(min=0, max=None)` and you can test for this condition and emit an `Activation` node in hls4ml (and bail otherwise until we support Clip in HLS).

The way converter handles `Gemm` and inserts a `Conv1D` if applied to 3D tensor is a wrong approach. What would happen if the test script did `Dense` after `Conv2D`? Your converter would fail. You should ensure that `Gemm` handler doesn't blindly assume that input is 2D and compute the correct output shape regardless of the input dimensionality. hls4ml now contains converters that will do the change from `Dense`->`Conv1D/2D` via an optimizer.

Those are my last two nitpicks :smile:

**Comment by Duchstf** [comment]
@vloncar fixed 😉

**Review by vloncar — APPROVED** [review]
I'll merge this now, but it needs to be followed up by cleanups

**Pull request merged**
